File: golang/two_pointers/remove_duplicates.go
package two_pointers

func removeDuplicates(arr []int) int {
	// time: O(n)
	// space: O(1)
	if len(arr) <= 1 {
		return len(arr)
	}
	
	// two pointers
	i, j := 0, 1
	
	for j < len(arr) {
		if arr[i] != arr[j] {
			i++
			arr[i] = arr[j]
		}
		
		j++
	}
	
	return i + 1
}

func removeAllKeys(arr []int, key int) int {
	// time: O(n)
	// space: O(1)
	i := 0
	
	for j := 0; j < len(arr); j++ {
		if arr[j] != key {
			arr[i] = arr[j]
			i++
		}
	}
	
	return i
}

File: javascript/src/two_pointers/removeDuplicates.js
export function removeDuplicates(arr) {
    // time: O(n)
    // space: O(1)
    if (arr.length <= 1) {
        return arr.length;
    }

    // two pointers
    let i = 0; // slow
    let j = 1; // fast

    while (j < arr.length) {
        if (arr[i] !== arr[j]) {
            i++;
            arr[i] = arr[j];
        }
        j++;
    }

    return i + 1;
}

export function removeAllKeys(arr, key) {
    // time: O(n)
    // space: O(1)
    let i = 0; // slow

    for (let j = 0; j < arr.length; j++) {
        if (arr[j] !== key) {
            arr[i] = arr[j];
            i++;
        }
    }

    return i;
}
